File: cpp/PrimeFactors/PrimeFactorsTest.cpp
#include "PrimeFactors.h"
#include <vector>
#include <gtest/gtest.h>
#include <stdarg.h>

void assertEquals(std::vector<int> expected, std::vector<int> actual) {
	std::vector<int>::iterator eIter, aIter;
	for (eIter = expected.begin(), aIter = actual.begin(); eIter != expected.end()
			&& aIter != actual.end(); eIter++, aIter++) {
		ASSERT_EQ(*eIter, *aIter);
	}

	std::vector<int>::size_type expectedSize = expected.size();
	std::vector<int>::size_type actualSize = actual.size();
	ASSERT_EQ(expectedSize, actualSize);
}

std::vector<int> emptyList(void) {
	std::vector<int> expected;
	return expected;
}

std::vector<int> asList(int one) {
	std::vector<int> expected = emptyList();
	expected.push_back(one);
	return expected;
}

TEST(PrimeFactorsTest, testOne)
{
	assertEquals(emptyList(), PrimeFactors::generate(1));
}

TEST(PrimeFactorsTest, testTwo)
{
	assertEquals(asList(2), PrimeFactors::generate(2));
}

TEST(PrimeFactorsTest, testThree)
{
	assertEquals(asList(3), PrimeFactors::generate(3));
}

TEST(PrimeFactorsTest, testFour)
{
	int input[2] = { 2, 2 };
	std::vector<int> expected(input, input + sizeof(input) / sizeof(int));

	assertEquals(expected, PrimeFactors::generate(4));
}

TEST(PrimeFactorsTest, testSix)
{
	int input[2] = { 2, 3 };
	std::vector<int> expected(input, input + sizeof(input) / sizeof(int));

	assertEquals(expected, PrimeFactors::generate(6));
}

TEST(PrimeFactorsTest, testEight)
{
	int input[3] = { 2, 2, 2 };
	std::vector<int> expected(input, input + sizeof(input) / sizeof(int));

	assertEquals(expected, PrimeFactors::generate(8));
}

TEST(PrimeFactorsTest, testNine)
{
	int input[2] = { 3, 3 };
	std::vector<int> expected(input, input + sizeof(input) / sizeof(int));

	assertEquals(expected, PrimeFactors::generate(9));
}

File: java/PrimeFactorsTest.java
package primefactors;

import static org.junit.Assert.*;
import static org.hamcrest.CoreMatchers.*;
import org.junit.Test;

import java.util.List;
import java.util.ArrayList;

public class PrimeFactorsTest {

  public static class PrimeFactors {
    public static List<Integer> generate(int n) {
      List<Integer> primes = new ArrayList<Integer>();
      int candidate = 2;
      while (n > 1) {
        while (n % candidate == 0) {
          primes.add(candidate);
          n /= candidate;
        }
        candidate++;
      }
      return primes;
    }
  }

  @Test
  public void testOne() {
    assertThat(PrimeFactors.generate(1), is(list()));
  }

  @Test
  public void testTwo() {
    assertThat(PrimeFactors.generate(2), is(list(2)));
  }

  @Test
  public void testThree() {
    assertThat(PrimeFactors.generate(3), is(list(3)));
  }

  @Test
  public void testFour() {
    assertThat(PrimeFactors.generate(4), is(list(2, 2)));
  }

  @Test
  public void testSix() {
    assertThat(PrimeFactors.generate(6), is(list(2, 3)));
  }

  @Test
  public void testEight() {
    assertThat(PrimeFactors.generate(8), is(list(2, 2, 2)));
  }

  @Test
  public void testNine() {
    assertThat(PrimeFactors.generate(9), is(list(3,3)));
  }

  public List<Integer> list(int... ints) {
    List<Integer> list = new ArrayList<Integer>();
    for (int i : ints) {
      list.add(i);
    }
    return list;
  }

}
